README

Parse::nm - Run and parse 'nm' command output with filter callbacks



INSTALLATION

To install this module, run the following commands:

        perl Build.PL
        ./Build
        ./Build test
        ./Build install


DEPENDENCIES

(see META.yml for up-to-date information)


SUPPORT AND DOCUMENTATION

After installing you can find documentation for this module with the
perldoc command:

        perldoc Parse::nm

You can also view the documentation of the latest release online:

    http://search.cpan.org/dist/Parse-nm/lib/Parse/nm.pm

Report bugs on the RT queue:

    http://rt.cpan.org/NoAuth/Bugs.html?Dist=Parse-nm


COPYRIGHT & LICENSE

Copyright © 2010-2011 Olivier Mengué

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.12.0 or,
at your option, any later version of Perl 5 you may have available.


Changes
Revision history for Parse::nm

0.09    2011-03-08T13:41+0100     DOLMEN (Olivier Mengué)
        Build.PL:
        - workaround for RT#66040: force Locale::Maketext
          1.17 dependency if an older one is present.
        - add configure_require dependency on Module::Build

0.08    2010-08-21T04:25+0200     DOLMEN (Olivier Mengué)
        Build.PL: Forgot Test::NoWarnings in build_depends.
        Now using Test::NoWarnings in all tests.

0.07    2010-08-21T04:00+0200     DOLMEN (Olivier Mengué)
        Fixed a bug when 'filters' are given in new() and not in parse().
        More tests. Test cover is at 91%.
        POD fixes. Added info about lack of support on OpenBSD.
        Windows still not supported.

0.06    2010-06-17T00:24+0200     DOLMEN (Olivier Mengué)
        Fixed ->run() on IRIX, thanks to David Cantrell.

0.05    2010-06-12T18:52+0200     DOLMEN (Olivier Mengué)
        Relax t/11-run.t to make it pass on Linux/Alpha, thanks to CPAN Testers.
        Relax t/11-run.t to make it pass on MacOS X, thanks to MADDINGUE.
        Released from « Les Journées Perl 2010 » in Calais.

0.04    2010-06-04T00:00+02:00    DOLMEN (Olivier Mengué)
        Full POD documentation.
        Testsuite, t/11-run.t:
        - fixed t/t.c to get more consistent behavior on different platforms.
        - now building using ExtUtils::CBuilder

0.03    2010-06-03T01:15+02:00    DOLMEN (Olivier Mengué)
        Added test 't/11-run.t' that compiles a C source and 'nm' it.
        Fixed minor issues.

0.02    2010-06-01T01:10+02:00    DOLMEN (Olivier Mengué)
        Added basic, incomplete, test suite.
        Use 3-args open() as suggested by Perl::Critic.
        Fixed other minor Perl::Critic suggestions.

0.01    2010-05-29T17:30+02:00    DOLMEN (Olivier Mengué)
        First basic release.
        No test suite.
        Basic, incomplete documentation.


t/t.c

char TestVar = 813;

int TestFunc()
{
    return 0;
}


lib/Parse/nm.pm
use strict;
use warnings;

package Parse::nm;

our $VERSION = '0.08';

use Carp 'croak';
use Regexp::Assemble;
use String::ShellQuote;

sub new
{
    my ($class, %args) = @_;
    _build_filters(\%args);
    return bless \%args, (ref $class ? ref $class : $class);
}

sub _build_filters
{
    my ($args) = @_;

    if (exists $args->{_comp_filters} && @{$args->{_comp_filters}}) {
	# Copy data to preserve $self
	$args->{_comp_filters} = [ @{$args->{_comp_filters}} ];
	$args->{_re} = $args->{_re}->clone;
    } else {
	$args->{_comp_filters} = [];
	$args->{_re} = Regexp::Assemble->new(fold_meta_pairs => 0);
    }

    if (exists $args->{filters}) {
        my @f = @{$args->{filters}};
        for my $f (@f) {
            my $name = $f->{name} || '\S+';
            my $type = $f->{type} || '[A-Z]';
            $args->{_re}->add("^$name +$type +");
            push @{$args->{_comp_filters}}, [
                qr/^($name) +($type) +/, $f->{action}
            ];
        }
	delete $args->{filters};
    }
}


sub run
{
    my ($self, %args) = @_;
    %args = (%{$self}, %args) if ref $self;

    my @options = exists $args{options} ? @{$args{options}} : ();
    my @files = ref $args{files} ? @{$args{files}} : ($args{files});

    my $nm;
    {
	# have to turn this on to get POSIX-ish output from nm -P on Irix
	local $ENV{_XPG} = '1' if ($^O eq 'irix');

	#open $nm, 'nm '.join(' ', map { my $x = $_; $x =~ s/"/\\"/g; qq{"$x"} } @files).' |'
	open $nm, '-|', shell_quote('nm', '-P', @options, @files)
	    or croak "Can't run 'nm': $!";
    }
    my $r = $self->parse($nm, %args);
    close $nm;
    return $r;
}


sub parse
{
    my ($self, $handle, %args) = @_;
    %args = (%{$self}, %args) if ref $self;
    _build_filters(\%args);
    my $re = $args{_re}->re;
    my $filters = $args{_comp_filters};
    while (<$handle>) {
        next unless /$re/;
        for my $f (@{$filters}) {
            if (/$f->[0]/) {
                $f->[1]($1, $2);
            }
        }
    }
    return ();
}

1;
__END__

=head1 NAME

Parse::nm - Run and parse 'nm' command output with filter callbacks

=head1 SYNOPSIS

Class interface:

    use Parse::nm;

    Parse::nm->run(options => [ qw(-e) ],
                   filters => [
                     {
                       name => qr/\.\w+/,
                       type => 'T',
                       action => sub {
                         print "$_[0]\n"
                       }
                     },
                   ],
                   files => 't.o',
                );

Object interface:

    use Parse::nm;

    my $pnm = Parse::nm->new(options => ...,
                             filters => ...);
    $pnm->run(files => 'file1.o');
    $pnm->run(files => 'file2.o');

    $str = "TestFunc T 0 0 \n";
    $pnm->parse(\$str);

=head1 METHODS

=head2 ->new(%options)

Builds an object with the default options.

=head2 ->parse(*GLOB, %options)

Parse 'nm -P'-style data coming from a filehandle.

Note that if your Perl is compiled with PerlIO (this is the default since
5.8.0), you can easily parse a string by opening a string reference to it.

    open($fh, '<', \$str);
    Parse::nm->parse($fh, %options);

=head2 ->run(%options)

Run C<nm> and parse its output.

=head1 OPTIONS

=over 4

=item options => [ ]

Command-line options given to C<nm> to run it.
The C<-P> (POSIX-style output) is always given.
C<-A> (show input file) is currently incompatible.

=item files => [ ]

List of files to give to C<nm> for parsing.

=item filters => \@filters

=over 4

=item name => qr/\S+/

A regexp that must match the name of the symbol.

Don't use C<^> or C<$>: this is not supported.

=item type => qr/[A-Z]/

A regexp that must match the type of the symbol. Types are single ASCII letter.
See the C<nm> man page of your operating system for more information.

Don't use C<^> or C<$>: this is not supported.

=item action => sub { ... }

A callback that will be triggered for each line where both C<name> and C<type>
match.

=back

=back

=head1 SEE ALSO

L<http://www.opengroup.org/onlinepubs/009695399/utilities/nm.html>

L<Binutils::Objdump>

=head1 PLATFORM SUPPORT

=over 4

=item POSIX systems

OK.

=item StrawberryPerl (Windows)

Work in progress (patches welcome).

=item OpenBSD

Parse::nm can not work on OpenBSD (at least up to 4.6) because 'nm' doesn't
have a POSIX-compatible mode.

L<http://www.openbsd.org/cgi-bin/man.cgi?query=nm&apropos=0&sektion=0&manpath=OpenBSD+4.6&arch=i386&format=html#STANDARDS>

=back

=head1 AUTHOR

Olivier MenguE<eacute>, C<dolmen@cpan.org>

=head1 COPYRIGHT AND LICENSE

Copyright E<copy> 2010 Olivier MenguE<eacute>.

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl itself, either Perl version 5.12.0 or, at your option,
any later version of Perl 5 you may have available.

=cut


Build.PL
#!/usr/bin/env perl
use utf8;  # vim:set ts=4 sw=4 sts=4:

use strict;
use warnings;

use Module::Build;

my $class = 'Module::Build';

if (-d '.git') {
	$class = $class->subclass(
		class => 'My::Builder',
		code => q{
			sub ACTION_tag {
                my ($self) = @_;
				my $v = $self->dist_version;
				system(qw(git tag -a -m), "CPAN release $v", "release-$v")
					or die "Git tagging failed.";
			}
		},
	);
}


# Conditional build_requires
my @build_requires;
# Workaround test problems when a broken version of Locale::Maketext is installed
# => we force an upgrade to 1.16+
# https://rt.cpan.org/Ticket/Display.html?id=66040
eval {
	use Locale::Maketext;
	#push @build_requires, 'Locale::Maketext' => '1.17' if Locale::Maketext->VERSION < '1.17';
};



my $builder = $class->new(
	module_name => 'Parse::nm',
	dist_author => 'Olivier Mengué <dolmen@cpan.org>',
	license     => 'perl',
	meta_add    => {
		resources => {
			homepage => 'http://github.com/dolmen/parse-nm',
			repository => 'http://github.com/dolmen/parse-nm.git',
			bugtracker => 'http://rt.cpan.org/NoAuth/Bugs.html?Dist=Parse-nm',
		},
	},
	requires    => {
		'perl' => '5.006',
		'Regexp::Assemble' => '0.34',
		'String::ShellQuote' => '1.04',
	},
	configure_requires => {
		'Module::Build' => '0.38',
	},
	build_requires => {
		'Module::Build' => '0.38',
		'Test::More' => '0.94',
		'File::Spec' => 0,
		'Test::NoWarnings' => '1.00',
		@build_requires,
	},
    build_recommends => {
        'ExtUtils::CBuilder' => '0.27',
    },
);
$builder->add_to_cleanup('MANIFEST.bak');
$builder->create_build_script();
